src/main/java/wolox/training/services/OpenLibrary.java

```diff
@@ -24,55 +29,94 @@
+        JsonObject json = JsonParser.parseString(jsonString).getAsJsonObject();
+        JsonObject jsonBook = json.get(String.format("ISBN:%s", isbn)).getAsJsonObject();
+
+        List<String> authors = new ArrayList<>();
+        jsonBook.get("authors").getAsJsonArray().forEach(i -> {
+            authors.add(i.getAsJsonObject().get("name").getAsString());
+        });
+
+        List<String> publishers = new ArrayList<>();
+        jsonBook.get("publishers").getAsJsonArray().forEach(i -> {
+            publishers.add(i.getAsJsonObject().get("name").getAsString());
+        });
+        JsonElement notes = jsonBook.get("notes");
+        String subtitle = "subtitle";
+        if (notes != null) {
+            subtitle = notes.getAsString();
+        }
+
+        JsonElement titleJson = jsonBook.get("title");
+        String title = "";
+        if (titleJson != null) {
+            title = titleJson.getAsString();
+        }
+
+        JsonElement publishDate = jsonBook.get("publish_date");
+        String published = "";
+        if (publishDate != null) {
+            published = publishDate.getAsString();
+        }
+
+        JsonElement pagesJson = jsonBook.get("number_of_pages");
+        int pages = 0;
+        if (pagesJson != null) {
+            pages = pagesJson.getAsInt();
+        }
+
+        return new OpenLibraryBookDTO(
+            isbn,
+            title,
+            subtitle,
+            publishers,
+            authors,
+            published,
+            pages
+        );
+    }
 
-    public OpenLibraryBookDTO tryGetBookByIsbn(String isbn) {
+    public Optional<Book> tryGetBookByIsbn(String isbn) {
         try {
-            JsonObject json = doIsbnRequest(isbn).getAsJsonObject();
-            JsonObject jsonBook = json.get(String.format("ISBN:%s", isbn)).getAsJsonObject();
-
-            List<String> authors = new ArrayList<>();
-            jsonBook.get("authors").getAsJsonArray().forEach(i -> {
-                authors.add(i.getAsJsonObject().get("name").getAsString());
-            });
-
-            List<String> publishers = new ArrayList<>();
-            jsonBook.get("publishers").getAsJsonArray().forEach(i -> {
-                publishers.add(i.getAsJsonObject().get("name").getAsString());
-            });
-
-            return new OpenLibraryBookDTO(
-                isbn,
-                jsonBook.get("title").getAsString(),
-                jsonBook.get("notes").getAsString(),
-                publishers,
-                authors,
-                jsonBook.get("publish_date").getAsString(),
-                jsonBook.get("number_of_pages").getAsInt()
-            );
-        } catch (Exception e) {
-            return null;
+            String jsonString = doIsbnRequest(isbn);
+            OpenLibraryBookDTO dto = buildDTOFromJsonString(isbn, jsonString);
+            return Optional.of(bookRepository.save(dto.getAsBook()));
+        } catch (IOException e) {
+            return Optional.empty();
```

Comment:
I think that if we use the `ExternalServiceException` we could let it to arise, because there is a problem with the external service indeed. 
Returning an Optional.empty here could be interpreted that there is no book with the given information

Reply:
:D